packages/hebao_v2/contracts/thirdparty/proxy/WalletProxy.sol

```diff
@@ -0,0 +1,48 @@
+// SPDX-License-Identifier: LGPL-3.0-or-later
+// Taken from: https://github.com/gnosis/safe-contracts/blob/development/contracts/proxies/GnosisSafeProxy.sol
+pragma solidity ^0.7.0;
+
+/// @title IProxy - Helper interface to access masterCopy of the Proxy on-chain
+/// @author Richard Meissner - <richard@gnosis.io>
+interface IProxy {
+    function masterCopy() external view returns (address);
+}
+
+/// @title WalletProxy - Generic proxy contract allows to execute all transactions applying the code of a master contract.
+/// @author Stefan George - <stefan@gnosis.io>
+/// @author Richard Meissner - <richard@gnosis.io>
+contract WalletProxy {
+
+    // masterCopy always needs to be first declared variable, to ensure that it is at the same location in the contracts to which calls are delegated.
+    // To reduce deployment costs this variable is internal and needs to be retrieved via `getStorageAt`
+    address internal masterCopy;
```

Comment:
Maybe we should make this masterCopy public so people can view and verify.

Reply:
@Brechtpd It doesn't seem this "masterCopy" is used anywhere in the code in a meaningful way. Could you please explain the idea, especailly compare to 'walletImplementation`?

Reply:
Oh, I get it. This is behind the WalletProxy contract.

Reply:
Yes, it's to make the upgrade of the wallet controlled by the wallet itself in a clear and efficient way (unlike `OwnedUpgradabilityProxy` which needs additional data for `owner` for example, which is always the wallet itself in this case).